src/OrleansRuntime/Core/InsideRuntimeClient.cs

```diff
@@ -364,35 +364,50 @@ public void SniffIncomingMessage(Message message)
                         throw exc;
                     }
 
-                    // If the target has a grain-level interceptor or there is a silo-level interceptor, intercept the call.
-                    var shouldCallSiloWideInterceptor = SiloProviderRuntime.Instance.GetInvokeInterceptor() != null && target is IGrain;
+                    Task<object> resultTask;
+
+                    // If the target has a grain-level interceptor or there is a silo-level interceptor, intercept the
+                    // call.
+                    var shouldCallSiloWideInterceptor = SiloProviderRuntime.Instance.GetInvokeInterceptor() != null
+                                                        && target is IGrain;
                     var intercepted = target as IGrainInvokeInterceptor;
-                    if (intercepted != null || shouldCallSiloWideInterceptor)
+                    var grainHasInterceptor = intercepted != null;
+                    if (grainHasInterceptor || shouldCallSiloWideInterceptor)
                     {
-                        // Fetch the method info for the intercepted call.
-                        var implementationInvoker =
-                            invocationMethodInfoMap.GetInterceptedMethodInvoker(target.GetType(), request.InterfaceId,
+                        // Get an invoker which delegates to the grain's IGrainInvocationInterceptor implementation.
+                        // If the grain does not implement IGrainInvocationInterceptor, then the invoker simply
+                        // delegates calls to the provided invoker.
+                        var interceptedInvoker =
+                            this.interceptedMethodInvokerCache.GetOrCreate(
+                                target.GetType(),
+                                request.InterfaceId,
                                 invoker);
-                        var methodInfo = implementationInvoker.GetMethodInfo(request.MethodId);
+                        var methodInfo = interceptedInvoker.GetMethodInfo(request.MethodId);
                         if (shouldCallSiloWideInterceptor)
                         {
                             // There is a silo-level interceptor and possibly a grain-level interceptor.
-                            var runtime = SiloProviderRuntime.Instance;
-                            resultObject =
-                                await runtime.CallInvokeInterceptor(methodInfo, request, target, implementationInvoker);
+                            // As a minor optimization, only pass the intercepted invoker if there is a grain-level
+                            // interceptor.
+                            resultTask = SiloProviderRuntime.Instance.CallInvokeInterceptor(
+                                methodInfo,
+                                request,
+                                target,
+                                grainHasInterceptor ? interceptedInvoker : invoker);
```

Comment:
My apologies for the confusion. I'll try to explain and then perhaps you can suggest a change.

* `invoker` is the plain-old (probably code-genned) `IGrainMethodInvoker`
* interceptedInvoker is an `IGrainMethodInvoker` which *wraps* `invoker`, passing all calls through `intercepted.Invoke` (note: `intercepted == target`, it's just cast to `IGrainInvokeInterceptor`)
* `SiloProviderRuntime.Instance.CallInvokeInterceptor` is the silo-wide interceptor.

We can have 4 cases, contingent on the presence of silo-level & grain-level interceptors.

 grain-level | silo-level | result | reason
---------------|------------|---------|---
NO |  NO | `invoker.Invoke()` | no interception at all
YES | NO  | `intercepted.Invoke()` | call the grain's `Invoke` method passing in `MethodInfo`
NO  | YES | `CallInvokeInterceptor(invoker)` | call the silo-level interceptor only
YES | YES | `CallInvokeInterceptor(interceptedInvoker)` | call the silo-level interceptor, passing an `IGrainMethodInvoker` which passes all calls through the grain-level interceptor

We use `intercepted` directly in the grain-level-only case instead of `interceptedInvoker` because the latter wraps the former to make it conform to the `IGrainMethodInvoker` interface and we don't need that wrapper. `intercepted` is the grain and implements `IGrainInvokeInterceptor` which has an identical method, `Invoke`, which takes one extra argument, `MethodInfo`.

Changes I could make (you may have more):
1. rename some of these variables & improve comments
2. split out the `MethodInfo` cache from `InterceptedMethodInvokerCache` so that there are two classes instead: one class which caches `MethodInfo`s and one which caches `InterceptedMethodInvoker`s
3. try to remove some if-statements/conditionals and always call a wrapped invoker (using a default one if no interceptor is specified at each level)

What are your thoughts?